### ISSUE TITLE
[Question]OC update scripts support

### ISSUE THREAD

**Issue by zjkyz8** [daliansky/XiaoMi-Pro-Hackintosh#518]
Is there any plan to provide OC update scripts? just like CLOVER.

**Comment by Menchen** [comment]
I'm working on an OC update script, it will be a 90% rewrite and mainly focused on OC update (no ALCPlug or other non-EFI patch) and I'm planning to support:
- [x] Select model to download from GitHub release (Need help, see below)
- [x] Backup old EFI
- [x] Install EFI  
- [x] Function to filter PList array with regex
- [ ] Enable BT SSDT based on old config
- [ ] Change DVMT/FBmem/platform-id based on old config
- [ ] Change 0xE2 patch based on old config
- [ ] Enable/Disable optional Kext based on old config (NullEthernel, intel wifi, intel Bluetooth...)
- [ ] PlatformInfo from old config
- [ ] Perseve mics preference like `ShowPicker` `Timeout`
- [ ] GTX support
- [ ] Vault support? not sure about this.

The download part is manual(It will ask the user about which model it should download) because there's is no good way to id the motherboard with OC? `bdmesg` is CLOVER only and for OC to debug info we need to use debug version of it... which is not recommended.
One solution is to get CPU info from macOS, but it is unreliable as it's faked.

The script is in a new file so the user can still install ALCPlugFix or use the non-CLOVER option from the clover script (I think we need to separate them into 2 scripts and source the install script from the main script)

**Comment by Menchen** [comment]
@stevezhengshiqi It's possible to add `BrcmBluetoothInjector.kext` `BrcmFirmwareData.kext` `BrcmPatchRAM3.kext` `AirportBrcmFixup.kext` to `makefile.sh` and `config.plist` and disable them by default?

I think downloading patch/Kext should be `makefile.sh`'s task and install/upgrade script which should only restore key value and not automating all user tasks.

What I have so far is https://github.com/Menchen/XiaoMi-Pro/blob/oc_install/Scripts/install_oc.sh, it's still a work in progress and untested.

The script should be able to restore Kext entry in plist with ease. See [this](https://github.com/Menchen/XiaoMi-Pro/blob/1e3b2a7936e7f430746749fae118c1272671bb7f/Scripts/install_oc.sh#L264), it loads the enabled value from the current config and set it to the new config. And thanks to OC we can enable/disable Kext without removing it from Kext folder.

And because it only restores, there's almost no chance of messing up the config file(it skip the restore in case of a missing entry due to OC config changing format), and we no need to explain what the option does(we have a wiki for that and it should match to the old configuration so no confirmation needed) with minimal interaction.

**Comment by stevezhengshiqi** [comment]
@Menchen Thanks for your work! I had some thoughts about OC scripts, but I barely have time to implement them.

> The download part is manual(It will ask the user about which model it should download) because there's is no good way to id the motherboard with OC? bdmesg is CLOVER only and for OC to debug info we need to use debug version of it... which is not recommended.

For this part, I plan to use ioreg | grep XHC's device-id to determine the user model. CML machine' XHC device-id is <ed020000> and KBL machine' device-id is <2f9d0000>. But users have to further choose GTX/MX350 after target EFI is downloaded.

> I think downloading patch/Kext should be makefile.sh's task and install/upgrade script which should only restore key value and not automating all user tasks.

Yeah, I agree with you. [download_kexts.sh](https://github.com/daliansky/XiaoMi-Pro-Hackintosh/blob/main/Kexts/download_kexts.sh) has full kext downloading functions and I can export them as APIs. However, these functions are not complex, and it may only take about 50 lines if you re-wheel in the install script. Either is fine.

> The script should be able to restore Kext entry in plist with ease. 

The restore function is great. But I still recommend allowing users' to choose if they change their ways to use BT. The restore function can be called if users choose "Keep same" or something like that.

I know the install script is like tons of tedious work, and feel free to implement it with no rush. I will ask @daliansky to give you collaborator privilege for you to handle potential conflicts or cross-script calls.

**Comment by stevezhengshiqi** [comment]
@Menchen 

> It's possible to add BrcmBluetoothInjector.kext BrcmFirmwareData.kext BrcmPatchRAM3.kext AirportBrcmFixup.kext to makefile.sh and config.plist and disable them by default?

I suggest to add them when running script using cat. Adding these entries would make config file overstaffed.

**Comment by Menchen** [comment]
@stevezhengshiqi 
>For this part, I plan to use ioreg | grep XHC's device-id to determine the user model. CML machine' XHC device-id is and KBL machine' device-id is <2f9d0000>. But users have to further choose GTX/MX350 after target EFI is downloaded.

The CML device-id is missing... Also, could you confirm 


```
ioreg -n XHC -c 'IOPCIDevice' -r -x -k "device-id" | perl -ne 'print $1 if(/^  \|   "device-id" = <((?:[0-9]|[a-f])*)>$/)'
```

 output the right device value?

> The restore function is great. But I still recommend allowing users' to choose if they change their ways to use BT. The restore function can be called if users choose "Keep same" or something like that.

I will keep that in mind... But not a priority right now as BT change should be rare...


> I suggest to add them when running script using cat. Adding these entries would make config file overstaffed.

If we're adding stuff to Plist I think the best way is to use `PlistBuddy` Import... But it needs a separate file.
And while we can embed the new Plist info in the script and write to a temp file but it complicates code, I think it's better to read from a folder the content of Plist as it's a resource and not "code".

The drawback would be the need to download the "plist patch" folder each time the script runs.
We can check if a file exists in Github using curl for each file and only download those needed. 
The script should be able to run in any folder without downloading the entire repo...

Another option is generating Kext entry on the fly... But it complicates code and things like `MinKernel` or `MaxKernel` won't be supported...

**Comment by stevezhengshiqi** [comment]
@Menchen 

> The CML device-id is missing... 

Sorry for that, it should be <ed020000>.

> Also, could you confirm

Nice job, it works like charm.

> If we're adding stuff to Plist I think the best way is to use PlistBuddy Import...

Although PlistBuddy is good at adding or editing a particular item in XML, I don't think it's convenient enough to handle multiple items. Like [one-key-hidpi](https://github.com/xzhih/one-key-hidpi/blob/master/hidpi.sh#L348-L360), the author uses cat to add a bunch of items.

I will check more details later, good effort so far.

**Comment by Menchen** [comment]
@stevezhengshiqi I have just finished the update script and it's ready for testing, I want your review before I push to the upstream.

**Disclaimer: This script assumes that you have a working and correct OC install, as it's only a tool to easily update to the last release!**

Use:
`bash <(curl -sL --proto-redir -all,https https://github.com/Menchen/XiaoMi-Pro/raw/oc_install/Scripts/install_oc.sh)`
Please use process substitution(`<()`) because piping the script to bash with curl break many things(and leave freezing bash behind).

There's a confirmation before deleting the existing EFI so it should be safe, also it does a backup in EFI partition, so it's easy to restore from a LiveUSB.

-------------------
I have disabled [2 functions](https://github.com/Menchen/XiaoMi-Pro/blob/d28be24736b4342013b8406a93bf16d7c316d9dc/Scripts/install_oc.sh#L918) because they download patches directly from the repo.

Those two functions download BrcmPatchRAM and AirportFixup from GitHub and download a patch, which is the list entry that needs to be inserted in `config.plist`, both functions are tested locally so there should be no problem.

---------------------------

One thing that needs confirmation is the version of `diff` that macOS came with.
```
❯ diff --version
diff (GNU diffutils) 2.8.1
Copyright (C) 2002 Free Software Foundation, Inc.

This program comes with NO WARRANTY, to the extent permitted by law.
You may redistribute copies of this program
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.

Written by Paul Eggert, Mike Haertel, David Hayes,
Richard Stallman, and Len Tower.
```
It's located in `/usr/bin/diff` which should be the "original" one from Apple... It's used to check if 2 folder is same so we don't create unnecessary backups.

---------------------------

I think the final script is better than what I thought, with a new function that **restores missing(user added) SSDT and Kext with a prompt** and another function that **finds Kext or SSDT that's enabled but with no file** with would cause OpenCore to abort.
Here's a small video featuring those features: https://asciinema.org/a/DVzc48BxjFwMCl14akKpoLpOl

With my config, the `NVMeFix.kext` was missing on the new release but the entry still exists in `config.plist`, and the script tries to restore it because it's an official [optional kext](https://github.com/Menchen/XiaoMi-Pro/blob/d28be24736b4342013b8406a93bf16d7c316d9dc/Scripts/install_oc.sh#L719)

Note that the function that restores SSDT/Kext copies the resource from existing EFI, so the `NVMeFix.kext` case can be considered a bug, but as long the user is aware I think it's ok to leave it as it. In the case that the user doesn't want to copy the missing recourse from the existing EFI, the script won't install and leave the user with an option to fix themself.

**Comment by stevezhengshiqi** [comment]
@Menchen Thank you so much for the script, it must take a lot of effort. After watching the video, I have to say that it's amazing. The script is powerful and clear.

I had a brief review on the code and looks good to me. I just have two suggestions at this stage.
 - In [L718](https://github.com/Menchen/XiaoMi-Pro/blob/oc_install/Scripts/install_oc.sh#L718) there're two NullEtherenet, maybe one of them is unnecessary?
 - Some indentations are not consistent in the script.

I will check more thoroughly after a while, and the script is good enough from my first view. Well done.

**Comment by Menchen** [comment]
> * In [L718](https://github.com/Menchen/XiaoMi-Pro/blob/oc_install/Scripts/install_oc.sh#L718), there's two NullEtherenet. Maybe one of them is unnecessary?

Indeed it's duplicated. I have fixed it, along with some more bugs and typos.

> * Some indentations are not consistent in the script.

It seems a problem with my editor with tab and space; I have fixed it with `beautysh`.


----------------------
While fixing bugs with optional kext that's missing in old EFI, so we need to disable them in new EFI, I have noticed in interactive restore the script only check for enabled SSDT/Kexts, should it also check missing/disabled entry in old EFI and ask the user whether they want to disable it? It's a rare user case.

Also, I have moved the backup right after the EFI mount; it should be fine as it detects duplicated backup and only creates when needed.

---------------------------------
Thanks for your reviewer, One thing I'm afraid that the script won't work on other machines due to external tool dependency.

I have opted `perl` in favor of `grep` because I have installed GNU Coreutils, and the BSD lacks many useful options.
The only tool that I'm not sure about is `diff` mentioned before, the only version I could found is the `GNU` version, so maybe it's not installed default in macOS.


--------------------------
I know reviewing near 1k lines of code is hard and time-consuming, so take your time with no rush.
